From 26bab1be400998807e9f83a088ff75134e42aa1a Mon Sep 17 00:00:00 2001
From: Furkan Kardame <f.kardame@manjaro.org>
Date: Wed, 15 Feb 2023 23:37:23 +0300
Subject: [PATCH] arm64: dts: rockchip: add dts for Orange PI 5 rk3588s

Add dts for Orange Pi 5.
Working IO:
* PMIC
* SD Card
* UART

Signed-off-by: Furkan Kardame <f.kardame@manjaro.org>
---
 arch/arm64/boot/dts/rockchip/Makefile         |   1 +
 .../boot/dts/rockchip/rk3588s-orangepi-5.dts  | 219 +++++++++++
 .../dts/rockchip/rk3588s-rk806-single.dtsi    | 350 ++++++++++++++++++
 3 files changed, 570 insertions(+)
 create mode 100644 arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
 create mode 100644 arch/arm64/boot/dts/rockchip/rk3588s-rk806-single.dtsi

diff --git a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
index 108cc298bd17b..e9e028392fde7 100644
--- a/arch/arm64/boot/dts/rockchip/Makefile
+++ b/arch/arm64/boot/dts/rockchip/Makefile
@@ -74,4 +74,5 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3568-evb1-v10.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3568-rock-3a.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-evb1-v10.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-rock-5b.dtb
+dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588s-orangepi-5.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588s-rock-5a.dtb
diff --git a/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts b/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
new file mode 100644
index 0000000000000..f9e38f004a87b
--- /dev/null
+++ b/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
@@ -0,0 +1,219 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+
+/dts-v1/;
+
+#include <dt-bindings/pinctrl/rockchip.h>
+#include <dt-bindings/gpio/gpio.h>
+#include "rk3588s.dtsi"
+#include "rk3588s-rk806-single.dtsi"
+
+/ {
+	model = "Orange Pi 5";
+	compatible = "rockchip,rk3588s-orange-pi-5", "rockchip,rk3588";
+		
+	chosen {
+		stdout-path = "serial2:1500000n8";
+	};
+	
+	vcc12v_dcin: vcc12v-dcin {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc12v_dcin";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <12000000>;
+		regulator-max-microvolt = <12000000>;
+	};
+
+	vcc5v0_sys: vcc5v0-sys {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0_sys";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		vin-supply = <&vcc12v_dcin>;
+	};
+
+	vcc5v0_usbdcin: vcc5v0-usbdcin {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0_usbdcin";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		vin-supply = <&vcc12v_dcin>;
+	};
+
+	vcc5v0_usb: vcc5v0-usb {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc5v0_usb";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		vin-supply = <&vcc5v0_usbdcin>;
+	};
+	
+	vcc_3v3_sd_s0: vcc-3v3-sd-s0 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc_3v3_sd_s0";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpios = <&gpio4 RK_PB5 GPIO_ACTIVE_LOW>;
+		enable-active-low;
+		vin-supply = <&vcc_3v3_s3>;
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-state-mem {
+			regulator-on-in-suspend;
+		};
+	};
+
+	vcc_1v1_nldo_s3: vcc-1v1-nldo-s3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vcc_1v1_nldo_s3";
+		regulator-always-on;
+		regulator-boot-on;
+		regulator-min-microvolt = <1100000>;
+		regulator-max-microvolt = <1100000>;
+		vin-supply = <&vcc5v0_sys>;
+	};
+	
+//		vbus5v0_typec: vbus5v0-typec {
+//		compatible = "regulator-fixed";
+//		regulator-name = "vbus5v0_typec";
+//		regulator-min-microvolt = <5000000>;
+//		regulator-max-microvolt = <5000000>;
+//		enable-active-high;
+//		gpio = <&gpio3 RK_PC0 GPIO_ACTIVE_HIGH>;
+//		vin-supply = <&vcc5v0_usb>;
+//		pinctrl-names = "default";
+//		pinctrl-0 = <&typec5v_pwren>;
+//	};
+
+//	vcc3v3_pcie2x1l2: vcc3v3-pcie2x1l2 {
+//		compatible = "regulator-fixed";
+//		regulator-name = "vcc3v3_pcie2x1l2";
+//		regulator-min-microvolt = <1800000>;
+//		regulator-max-microvolt = <1800000>;
+//		enable-active-high;
+//		regulator-boot-on;
+//		regulator-always-on;
+//		gpios = <&gpio0 RK_PC5 GPIO_ACTIVE_HIGH>;
+//		startup-delay-us = <50000>;
+//		vin-supply = <&vcc5v0_sys>;
+//	};
+};
+
+&cpu_l0 {
+	cpu-supply = <&vdd_cpu_lit_s0>;
+	mem-supply = <&vdd_cpu_lit_s0>;
+};
+
+&cpu_b0 {
+	cpu-supply = <&vdd_cpu_big0_s0>;
+	mem-supply = <&vdd_cpu_big0_mem_s0>;
+	status = "disabled";
+};
+
+&cpu_b2 {
+	cpu-supply = <&vdd_cpu_big1_s0>;
+	mem-supply = <&vdd_cpu_big1_mem_s0>;
+	status = "disabled";
+};
+
+&i2c0 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c0m2_xfer>;
+
+	vdd_cpu_big0_s0: vdd_cpu_big0_mem_s0: rk8602@42 {
+		compatible = "rockchip,rk8602";
+		reg = <0x42>;
+		vin-supply = <&vcc5v0_sys>;
+		regulator-compatible = "rk860x-reg";
+		regulator-name = "vdd_cpu_big0_s0";
+		regulator-min-microvolt = <550000>;
+		regulator-max-microvolt = <1050000>;
+		regulator-ramp-delay = <2300>;
+		rockchip,suspend-voltage-selector = <1>;
+		regulator-boot-on;
+		regulator-always-on;
+		regulator-state-mem {
+			regulator-off-in-suspend;
+		};
+	};
+
+	vdd_cpu_big1_s0: vdd_cpu_big1_mem_s0: rk8603@43 {
+		compatible = "rockchip,rk8603";
+		reg = <0x43>;
+		vin-supply = <&vcc5v0_sys>;
+		regulator-compatible = "rk860x-reg";
+		regulator-name = "vdd_cpu_big1_s0";
+		regulator-min-microvolt = <550000>;
+		regulator-max-microvolt = <1050000>;
+		regulator-ramp-delay = <2300>;
+		rockchip,suspend-voltage-selector = <1>;
+		regulator-boot-on;
+		regulator-always-on;
+		regulator-state-mem {
+			regulator-off-in-suspend;
+		};
+	};
+};
+
+&i2c2 {
+	status = "disabled";
+
+	vdd_npu_mem_s0: rk8602@42 {
+		compatible = "rockchip,rk8602";
+		reg = <0x42>;
+		vin-supply = <&vcc5v0_sys>;
+		regulator-compatible = "rk860x-reg";
+		regulator-name = "vdd_npu_s0";
+		regulator-min-microvolt = <550000>;
+		regulator-max-microvolt = <950000>;
+		regulator-ramp-delay = <2300>;
+		rockchip,suspend-voltage-selector = <1>;
+		regulator-boot-on;
+		regulator-always-on;
+		regulator-state-mem {
+			regulator-off-in-suspend;
+		};
+	};
+};
+
+&sdhci {
+	bus-width = <8>;
+	no-sdio;
+	no-sd;
+	non-removable;
+	max-frequency = <200000000>;
+	mmc-hs400-1_8v;
+	mmc-hs400-enhanced-strobe;
+	status = "disabled";
+};
+
+&sdmmc {
+	max-frequency = <150000000>;
+	no-sdio;
+	no-mmc;
+	bus-width = <4>;
+	cap-mmc-highspeed;
+	cap-sd-highspeed;
+	disable-wp;
+	sd-uhs-sdr104;
+	vmmc-supply = <&vcc_3v3_sd_s0>;
+	vqmmc-supply = <&vccio_sd_s0>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2m0_xfer>;
+	status = "okay";
+};
+
+&tsadc {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/rockchip/rk3588s-rk806-single.dtsi b/arch/arm64/boot/dts/rockchip/rk3588s-rk806-single.dtsi
new file mode 100644
index 0000000000000..f66db3b3f1ba4
--- /dev/null
+++ b/arch/arm64/boot/dts/rockchip/rk3588s-rk806-single.dtsi
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2021 Rockchip Electronics Co., Ltd.
+ *
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+
+&spi2 {
+        status = "okay";
+        assigned-clocks = <&cru CLK_SPI2>;
+        assigned-clock-rates = <200000000>;
+        num-cs = <2>;
+
+        rk806master: pmic@0 {
+                compatible = "rockchip,rk806";
+                spi-max-frequency = <1000000>;
+                reg = <0x0>;
+
+                interrupt-parent = <&gpio0>;
+                interrupts = <7 IRQ_TYPE_LEVEL_LOW>;
+
+                pinctrl-names = "default", "pmic-power-off";
+                pinctrl-0 = <&pmic_pins>, <&rk806_dvs1_null>, <&rk806_dvs2_null>, <&rk806_dvs3_null>;
+                pinctrl-1 = <&rk806_dvs1_pwrdn>;
+
+                vcc1-supply = <&vcc5v0_sys>;
+                vcc2-supply = <&vcc5v0_sys>;
+                vcc3-supply = <&vcc5v0_sys>;
+                vcc4-supply = <&vcc5v0_sys>;
+                vcc5-supply = <&vcc5v0_sys>;
+                vcc6-supply = <&vcc5v0_sys>;
+                vcc7-supply = <&vcc5v0_sys>;
+                vcc8-supply = <&vcc5v0_sys>;
+                vcc9-supply = <&vcc5v0_sys>;
+                vcc10-supply = <&vcc5v0_sys>;
+                vcc11-supply = <&vcc_2v0_pldo_s3>;
+                vcc12-supply = <&vcc5v0_sys>;
+                vcc13-supply = <&vcc_1v1_nldo_s3>;
+                vcc14-supply = <&vcc_1v1_nldo_s3>;
+                vcca-supply = <&vcc5v0_sys>;
+
+                gpio-controller;
+                #gpio-cells = <2>;
+
+                pwrkey {
+                        status = "okay";
+                };
+
+				rk806_dvs1_null: dvs1-null-pins {
+                        pins = "gpio_pwrctrl1";
+                        function = "pin_fun0";
+                };
+
+                rk806_dvs1_slp: dvs1-slp-pins {
+                        pins = "gpio_pwrctrl1";
+                        function = "pin_fun1";
+                };
+
+                rk806_dvs1_pwrdn: dvs1-pwrdn-pins {
+                        pins = "gpio_pwrctrl1";
+                        function = "pin_fun2";
+                };
+
+                rk806_dvs1_rst: dvs1-rst-pins {
+                        pins = "gpio_pwrctrl1";
+                        function = "pin_fun3";
+                };
+
+                rk806_dvs2_null: dvs2-null-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun0";
+                };
+
+                rk806_dvs2_slp: dvs2-slp-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun1";
+                };
+
+                rk806_dvs2_pwrdn: dvs2-pwrdn-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun2";
+                };
+
+                rk806_dvs2_rst: dvs2-rst-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun3";
+                };
+
+                rk806_dvs2_dvs: dvs2-dvs-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun4";
+                };
+
+                rk806_dvs2_gpio: dvs2-gpio-pins {
+                        pins = "gpio_pwrctrl2";
+                        function = "pin_fun5";
+                };
+
+                rk806_dvs3_null: dvs3-null-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun0";
+                };
+
+                rk806_dvs3_slp: dvs3-slp-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun1";
+                };
+
+                rk806_dvs3_pwrdn: dvs3-pwrdn-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun2";
+                };
+
+                rk806_dvs3_rst: dvs3-rst-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun3";
+                };
+
+                rk806_dvs3_dvs: dvs3-dvs-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun4";
+                };
+
+                rk806_dvs3_gpio: dvs3-gpio-pins {
+                        pins = "gpio_pwrctrl3";
+                        function = "pin_fun5";
+                };
+
+                regulators {
+                        vdd_gpu_s0: dcdc-reg1 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <550000>;
+                                regulator-max-microvolt = <950000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_gpu_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd_cpu_lit_s0: dcdc-reg2 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <550000>;
+                                regulator-max-microvolt = <950000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_cpu_lit_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd_log_s0: dcdc-reg3 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <675000>;
+                                regulator-max-microvolt = <750000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_log_s0";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                        regulator-suspend-microvolt = <750000>;
+                                };
+                        };
+
+                        vdd_vdenc_s0: dcdc-reg4 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <550000>;
+                                regulator-max-microvolt = <950000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_vdenc_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd_gpu_mem_s0: dcdc-reg5 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <675000>;
+                                regulator-max-microvolt = <950000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_gpu_mem_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd2_ddr_s3: dcdc-reg6 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-name = "vdd2_ddr_s3";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                };
+                        };
+
+                        vcc_2v0_pldo_s3: dcdc-reg7 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <2000000>;
+                                regulator-max-microvolt = <2000000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vdd_2v0_pldo_s3";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                        regulator-suspend-microvolt = <2000000>;
+                                };
+                        };
+
+                        vcc_3v3_s3: dcdc-reg8 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <3300000>;
+                                regulator-max-microvolt = <3300000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "vcc_3v3_s3";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                        regulator-suspend-microvolt = <3300000>;
+                                };
+                        };
+
+                        vddq_ddr_s0: dcdc-reg9 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-name = "vddq_ddr_s0";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                };
+                        };
+
+                        vdd_ddr_s0: dcdc-reg10 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-name = "vdd_ddr_s0";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                };
+                        };
+
+                        avcc_1v8_s0: pldo-reg1 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <1800000>;
+                                regulator-max-microvolt = <1800000>;
+                                regulator-ramp-delay = <12500>;
+                                regulator-name = "avcc_1v8_s0";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                };
+                        };
+
+                        vdd1_1v8_ddr_s3: pldo-reg2 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <900000>;
+                                regulator-max-microvolt = <900000>;
+                                regulator-name = "vdd2l_0v9_ddr_s3";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                        regulator-suspend-microvolt = <900000>;
+                                };
+                        };
+
+                        vcc_3v3_s0: pldo-reg4 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <3300000>;
+                                regulator-max-microvolt = <3300000>;
+                                regulator-name = "vcc_3v3_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vccio_sd_s0: pldo-reg5 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <1800000>;
+                                regulator-max-microvolt = <3300000>;
+                                regulator-name = "vccio_sd_s0";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                };
+                        };
+
+                        vdd_0v75_s3: NLDO_REG1 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <750000>;
+                                regulator-max-microvolt = <750000>;
+                                regulator-name = "vdd_0v75_s3";
+                                regulator-state-mem {
+                                        regulator-on-in-suspend;
+                                        regulator-suspend-microvolt = <750000>;
+                                };
+                        };
+
+                        avdd_0v75_s0: nldo-reg3 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <750000>;
+                                regulator-max-microvolt = <750000>;
+                                regulator-name = "avdd_0v75_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd_ddr_pll_s0: nldo_reg2 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <850000>;
+                                regulator-max-microvolt = <850000>;
+                                regulator-name = "vdd_ddr_pll_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                        regulator-suspend-microvolt = <850000>;
+                                };
+                        };
+
+                        vdd_0v85_s0: nldo-reg4 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <850000>;
+                                regulator-max-microvolt = <850000>;
+                                regulator-name = "vdd_0v85_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+
+                        vdd_0v75_hdmi_edp_s0: nldo-reg5 {
+                                regulator-always-on;
+                                regulator-boot-on;
+                                regulator-min-microvolt = <750000>;
+                                regulator-max-microvolt = <750000>;
+                                regulator-name = "vdd_0v75_hdmi_edp_s0";
+                                regulator-state-mem {
+                                        regulator-off-in-suspend;
+                                };
+                        };
+                };
+        };
+};
